tcs-client/src/test/java/com/transformuk/hee/tis/tcs/client/service/impl/TcsServiceImplTest.java

```diff
@@ -240,6 +249,30 @@ public void getSpecialtyByIdShouldFindSpecialtyDto() {
     verify(restTemplate).getForEntity(url, SpecialtyDTO.class);
   }
 
+  @Test
+  public void getSpecialtyByNameShouldFindSpecialtyDto() throws EncoderException {
+    SpecialtyDTO specialty = new SpecialtyDTO();
+    specialty.setName("specialtyName");
+    specialty.setSpecialtyTypes(new HashSet<>(
+        Collections.singletonList(SpecialtyType.SUB_SPECIALTY)));
+    specialty.setId(20L);
+
+    String encodedParameters = new URLCodec().encode("{\"name\":[\"specialtyName\"],"
+        + "\"status\":[\"CURRENT\"],\"specialtyTypes\":[\"SUB_SPECIALTY\"]}");
+    String url = "http://localhost:9999/tcs/api/specialties?columnFilters=" + encodedParameters;
+
+    List<SpecialtyDTO> expectedResult = Arrays.asList(specialty);
+    ResponseEntity responseEntity = new ResponseEntity(expectedResult, HttpStatus.OK);
+    doReturn(responseEntity).when(restTemplate).exchange(url, HttpMethod.GET, null,
+        new ParameterizedTypeReference<List<SpecialtyDTO>>() {});
+
+    List<SpecialtyDTO> result = testObj.getSpecialtyByName("specialtyName",
+        SpecialtyType.SUB_SPECIALTY);
+    assertThat("Unexpected result", result, is(expectedResult));
+    verify(restTemplate).exchange(url, HttpMethod.GET, null,
+        new ParameterizedTypeReference<List<SpecialtyDTO>>() {});
+  }
+
```

Comment:
It would be nice to test the opposite condition, i.e. that it doesn't find specialties by name if the type doesn't match. Maybe something like:

```
@Test
  public void getSpecialtyByNameShouldNotFindDifferentTypeSpecialtyDto() throws EncoderException {
    SpecialtyDTO specialty = new SpecialtyDTO();
    specialty.setName("specialtyName");
    specialty.setSpecialtyTypes(new HashSet<>(
        Collections.singletonList(SpecialtyType.SUB_SPECIALTY)));
    specialty.setId(20L);

    String encodedParameters = new URLCodec().encode("{\"name\":[\"specialtyName\"],"
        + "\"status\":[\"CURRENT\"],\"specialtyTypes\":[\"PLACEMENT\"]}");
    String url = "http://localhost:9999/tcs/api/specialties?columnFilters=" + encodedParameters;

    List<SpecialtyDTO> expectedResult = Lists.emptyList();
    ResponseEntity responseEntity = new ResponseEntity(expectedResult, HttpStatus.OK);
    doReturn(responseEntity).when(restTemplate).exchange(url, HttpMethod.GET, null,
        new ParameterizedTypeReference<List<SpecialtyDTO>>() {});
    
    exceptionRule.expect(HttpClientErrorException.class);
    exceptionRule.expectMessage("404 Not Found");

    List<SpecialtyDTO> result = testObj.getSpecialtyByName("specialtyName",
        SpecialtyType.SUB_SPECIALTY);
    assertThat("Unexpected result", result, is(expectedResult));
    verify(restTemplate).exchange(url, HttpMethod.GET, null,
        new ParameterizedTypeReference<List<SpecialtyDTO>>() {});
  }
```

Otherwise it all looks good! :)

Reply:
This test passes but I think it's for the wrong reasons, which is `restTemplate.exchange(<URL for SUB_SPECIALTY> etc...)` not being stubbed. The url for retrieving PLACEMENT types of Specialties is formulated but it's never used in the test, which still only calls `testObj.getSpecialtyByName("specialtyName", SpecialtyType.SUB_SPECIALTY)`, and therefore it'll try to use `restTemplate.exchange(<URL for SUB_SPECIALTY> etc...)` instead of `restTemplate.exchange(<URL for PLACEMENT> etc...)`.

The actual functionality under test here would be that given a name and specialty type to be passed to the method, when the method is called, then it results in a REST GET call with such parameters. Everything happening afterwards is responsibility of `restTemplate` and of the endpoint, which is tested elsewhere. Also I wouldn't expect the call to raise an exception with a 404, since it's looking for matching dtos, so I'd expect a 200 with an empty list.

I added this test and tweaked it so that it stubs both scenarios (SUB_SPECIALTY and non-SUB_SPECIALTY), and verifies that the resulting REST call is consistent with the argument being passed to the method and isn't with the other argument passed to the method.

Reply:
Ah, that's true - I was being hasty.